### PR TITLE
support SQL-based UCR rebuilds

### PR DESCRIPTION
**Pull request by czue**
probably easiest :fish: though one commit was a bit of a wrong turn.

@snopoke cc @calellowitz @millerdev

### CODE REVIEW

**Comment by millerdev** [comment]
> probably easiest :fish: though one commit was a bit of a wrong turn.

what does [:fish:](https://vulcanpost.com/7721/why-i-hate-emoji-icons-and-why-you-probably-should-too/) mean in this context?

**Comment by czue** [comment]
lol, sorry, it means commit-wise (as opposed to :office: which is all at once)

**Comment by millerdev** [comment]
haha, how did those icons come to mean those things?

not sure if you saw the link in my previous post, which doesn't quite describe my dislike for emoji perfectly, but got the fish part right. I often find emoji to be a kind of in-crowd-cool-kids phenomenon, and more often than not I find myself wondering if I understood what the original author was trying to say. Writing sentences with emoji is like coding with single-letter variable names: it's fast and fun, but hard to understand when you have to go back and read it.

**Comment by czue** [comment]
I think I missed the link, though I do agree that it's bad when emoji usage is used to convey meaning when not everyone would be able to understand the meaning.

The :fish: / :office: thing came out of the scale team (the etymology is commit by commit --> blow by blow --> :blowfish: / :blowfish: --> :blowfish: --> :fish: and now anything that lives in the sea is fair game for commit-wise). Agree we shouldn't use it if not everyone is on board, but I think like :+1: and :white_check_mark: it's a concise (and fun) way to convey something useful as long as everyone knows what it means. And hopefully with the review buddy system that keeps happening.

File: corehq/ex-submodules/couchforms/dbaccessors.py [hunk 1]
```diff
@@ -30,6 +30,20 @@ def get_form_ids_by_type(domain, type_, start=None, end=None):
     )]
 
 
+def get_form_ids_by_type(domain, type_):
```
Comment (dimagimon) [inline corehq/ex-submodules/couchforms/dbaccessors.py]: F811 redefinition of unused 'get_form_ids_by_type' from line 14

**Comment by millerdev** [comment]
Maybe we should add a PR emoji definitions to our code review guide? I've been wondering what :fish: and :office: mean for a long time. Thanks for explaining.

File: corehq/ex-submodules/couchforms/dbaccessors.py (hunk 1 above)
Comment (millerdev) [inline corehq/ex-submodules/couchforms/dbaccessors.py]: this is almost certainly a bug

**Comment by benrudolph** [comment]
good points, i think it makes sense to add them to the code review guidelines as an option. i think those are the few emojis that have meaning. in general i take a string of emojis as a fun puzzle and if i can't figure it out i'm not missing anything essential and on the flip side i never try to convey meaningful feedback with emojis

File: corehq/ex-submodules/couchforms/dbaccessors.py (hunk 1 above)
Comment (czue) [inline corehq/ex-submodules/couchforms/dbaccessors.py]: wow, i just completely rewrote a function that already existed. :disappointed:

**Comment by czue** [comment]
updated to include the form-reindexes too. should be g2g once :white_check_mark:

**Comment by millerdev** [comment]
good to merge on test pass

**Comment by czue** [comment]
:white_check_mark:

**Comment by czue** [comment]
:white_check_mark:

File: corehq/sql_accessors/sql_templates/get_form_ids_in_domain_by_type.sql [hunk 2]
```diff
@@ -0,0 +1,11 @@
+-- drop it first in case we're changing the signature in which case 'CREATE OR REPLACE' will fail
+DROP FUNCTION IF EXISTS get_form_ids_in_domain_by_type(TEXT, INTEGER);
+
+-- return SETOF so that we get 0 rows when no form matches otherwise we'll get an empty row
+CREATE FUNCTION get_form_ids_in_domain_by_type(domain_name TEXT, state INTEGER DEFAULT 1) RETURNS TABLE (form_id VARCHAR(255)) AS $$
+BEGIN
+    RETURN QUERY  EXECUTE
+    'SELECT form_table.form_id FROM form_processor_xforminstancesql as form_table WHERE form_table.domain = $1 AND form_table.state = $2'
+    USING domain_name, state;
```
Comment (snopoke) [inline corehq/sql_accessors/sql_templates/get_form_ids_in_domain_by_type.sql]: This doesn't need to use execute. You can just run the query normally (but you may have to change the name of the `state` param):
```
RETURN QUERY
SELECT form_table.form_id FROM form_processor_xforminstancesql as form_table WHERE form_table.domain = domain_name AND form_table.state = p_state;
```

File: corehq/form_processor/backends/sql/dbaccessors.py [hunk 3]
```diff
@@ -269,6 +269,17 @@ def get_deleted_form_ids_for_user(domain, user_id):
...
     @staticmethod
+    def get_form_ids_in_domain_by_type(domain, type_):
+        state = doc_type_to_state[type_]
+        with get_cursor(XFormInstanceSQL) as cursor:
+            cursor.execute(
+                'SELECT form_id from get_form_ids_in_domain_by_type(%s, %s)',
+                [domain, state]
+            )
+            results = fetchall_as_namedtuple(cursor)
+            return [result.form_id for result in results]
```
Comment (snopoke) [inline corehq/form_processor/backends/sql/dbaccessors.py]: Do you think we should make `fetchall_as_namedtuple` a generator so and then also `yeild` from this function?

Comment (czue) [inline corehq/form_processor/backends/sql/dbaccessors.py]: Yeah that seems good. Might be a larger change though (looks like lots of dbaccessors are using it)

File: corehq/sql_accessors/sql_templates/get_form_ids_in_domain_by_type.sql (hunk 2 above)
Comment (czue) [inline corehq/sql_accessors/sql_templates/get_form_ids_in_domain_by_type.sql]: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/userreports/README.md#nested-expressions

**Comment by snopoke** [comment]
:+1:  on :white_check_mark: but feel free to merge if you tested locally

**Comment by czue** [comment]
:white_check_mark:

**Pull request merged**